pkg/controllers/cinder/cinder.go

```diff
@@ -0,0 +1,151 @@
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	"github.com/openshift/openstack-cinder-csi-driver-operator/pkg/util"
+	corev1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/kubernetes"
+	storagelisters "k8s.io/client-go/listers/storage/v1"
+	"k8s.io/klog/v2"
+)
+
+// This CinderController watches OpenStack and:
+// 1) Installs Cinder CSI driver itself.
+// 2) Creates StorageClass for each volume type provided by Cinder.
+//
+// Note that StorageClasses are not deleted when a volumes type disappears
+// from Cinder.
+
+type CinderController struct {
+	operatorClient     v1helpers.OperatorClient
+	kubeClient         kubernetes.Interface
+	openStackClient    *openStackClient
+	storageClassLister storagelisters.StorageClassLister
+	// Controllers to start when Cinder is detected
+	csiControllers     []Runnable
+	controllersRunning bool
+	eventRecorder      events.Recorder
+}
+
+type Runnable interface {
+	Run(ctx context.Context, workers int)
+}
+
+const (
+	// Minimal interval between controller resyncs. The controller will detect
+	// new share types in Cinder and create StorageClasses for them at least
+	// once per this interval.
+	resyncInterval = 20 * time.Minute
+
+	operatorConditionPrefix = "CinderController"
+)
+
+func NewCinderController(
+	operatorClient v1helpers.OperatorClient,
+	kubeClient kubernetes.Interface,
+	informers v1helpers.KubeInformersForNamespaces,
+	openStackClient *openStackClient,
+	csiControllers []Runnable,
+	eventRecorder events.Recorder) factory.Controller {
+
+	scInformer := informers.InformersFor("").Storage().V1().StorageClasses()
+	c := &CinderController{
+		operatorClient:     operatorClient,
+		kubeClient:         kubeClient,
+		storageClassLister: scInformer.Lister(),
+		openStackClient:    openStackClient,
+		csiControllers:     csiControllers,
+		eventRecorder:      eventRecorder.WithComponentSuffix("CinderController"),
+	}
+	return factory.New().WithSync(c.sync).WithSyncDegradedOnError(operatorClient).ResyncEvery(resyncInterval).WithInformers(
+		operatorClient.Informer(),
+		scInformer.Informer(),
+	).ToController("CinderController", eventRecorder)
+}
+
+func (c *CinderController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	klog.V(4).Infof("Cinder sync started")
+	defer klog.V(4).Infof("Cinder sync finished")
+
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorv1.Managed {
+		return nil
+	}
+
+	volumeTypes, err := c.openStackClient.GetVolumeTypes()
+	if err != nil {
+		return err
+	}
+
+	if len(volumeTypes) == 0 {
+		err = fmt.Errorf("no volume types found for Cinder service")
+		klog.V(2).Error(err, "No volume types found for Cinder service")
+		return err
```

Comment:
Returning an error here means that the operator will get Degraded (see `WithSyncDegradedOnError()` above).

I don't think it's reasonable to do that if `c.openStackClient.GetVolumeTypes()` returns an error or an empty slice. Perhaps would it be OK to just not create the StorageClasses?